ci-operator/step-registry/ipi/conf/aws/user-min-permissions/ipi-conf-aws-user-min-permissions-commands.sh

```diff
@@ -0,0 +1,282 @@
+iam:ListRoles
+iam:ListUserPolicies
+iam:ListUsers
+iam:PassRole
+iam:PutRolePolicy
+iam:PutUserPolicy
+iam:RemoveRoleFromInstanceProfile
+iam:SimulatePrincipalPolicy
+iam:TagRole
+iam:TagUser
+iam:UntagRole
+route53:ChangeResourceRecordSets
+route53:ChangeTagsForResource
+route53:CreateHostedZone
+route53:DeleteHostedZone
+route53:GetChange
+route53:GetHostedZone
+route53:ListHostedZones
+route53:ListHostedZonesByName
+route53:ListResourceRecordSets
+route53:ListTagsForResource
+route53:UpdateHostedZoneComment
+s3:AbortMultipartUpload
+s3:CreateBucket
+s3:DeleteBucket
+s3:DeleteObject
+s3:GetAccelerateConfiguration
+s3:GetBucketAcl
+s3:GetBucketCors
+s3:GetBucketLocation
+s3:GetBucketLogging
+s3:GetBucketObjectLockConfiguration
+s3:GetBucketPublicAccessBlock
+s3:GetBucketReplication
+s3:GetBucketRequestPayment
+s3:GetBucketTagging
+s3:GetBucketVersioning
+s3:GetBucketWebsite
+s3:GetEncryptionConfiguration
+s3:GetLifecycleConfiguration
+s3:GetObject
+s3:GetObjectAcl
+s3:GetObjectTagging
+s3:GetObjectVersion
+s3:GetReplicationConfiguration
+s3:HeadBucket
+s3:ListBucket
+s3:ListBucketMultipartUploads
+s3:ListBucketVersions
+s3:PutBucketAcl
+s3:PutBucketPublicAccessBlock
+s3:PutBucketTagging
+s3:PutEncryptionConfiguration
+s3:PutLifecycleConfiguration
+s3:PutObject
+s3:PutObjectAcl
+s3:PutObjectTagging
+servicequotas:ListAWSDefaultServiceQuotas
+tag:GetResources
+EOF
+	# additional permisions for 4.11+
+	if ((ocp_minor_version >= 11 && ocp_major_version == 4)); then
+		echo "ec2:DeletePlacementGroup" >>"${PERMISION_LIST}"
+		echo "s3:GetBucketPolicy" >>"${PERMISION_LIST}"
+	fi
+
+	# additional permisions for 4.14+
+	if ((ocp_minor_version >= 14 && ocp_major_version == 4)); then
+		echo "ec2:DescribeSecurityGroupRules" >>"${PERMISION_LIST}"
+
+		# sts:AssumeRole is required for Shared-VPC install https://issues.redhat.com/browse/OCPBUGS-17751
+		echo "sts:AssumeRole" >>"${PERMISION_LIST}"
+	fi
+
+	# additional permisions for 4.15+
+	if ((ocp_minor_version >= 15 && ocp_major_version == 4)); then
+		echo "iam:TagInstanceProfile" >>"${PERMISION_LIST}"
+	fi
+
+	# additional permisions for 4.16+
+	if ((ocp_minor_version >= 16 && ocp_major_version == 4)); then
+		echo "ec2:DisassociateAddress" >>"${PERMISION_LIST}"
+		echo "elasticloadbalancing:SetSecurityGroups" >>"${PERMISION_LIST}"
+		echo "s3:PutBucketPolicy" >>"${PERMISION_LIST}"
+	fi
+
+	cat <<EOF >"/tmp/jsoner.py"
+import json
+import sys
+
+data = sys.stdin.read().splitlines()
+out = {
+	"Version": "2012-10-17",
+	"Statement": [{
+		"Effect": "Allow",
+		"Resource": "*",
+		"Action": data,
+	}]
+}
+print(json.dumps(out, indent=2))
+EOF
+	# generate policy file
+	cat "${PERMISION_LIST}" | python3 /tmp/jsoner.py >"${USER_POLICY_FILE}"
+	rm -f /tmp/jsoner.py
+
+else
+	dir=/tmp/min_perms/
+
+	mkdir -p ${dir}
+
+	# Make a copy of the install-config.yaml since the installer will consume it.
+	cp "${SHARED_DIR}/install-config.yaml" ${dir}/
+
+	openshift-install create permissions-policy --dir ${dir}
```

Comment:
@r4f4  Installer will generate various permission depending on different install-config [1], so if testing byo-iamProfile feature with minimum permission, we need to invoke `ref: ipi-conf-aws-byo-iam-profile` before `chain: ipi-conf-aws`, is that right?

[1] https://github.com/openshift/installer/pull/8704/files#diff-d16efb382a781ecc705d5d81ff99bf29e0333b7abce40bd409a1199c16099c74R47

Reply:
yes, any modifications to the `install-config.yaml` need to happen before `create permissions-policy` is invoked. I cannot add the min perms steps to the start of the `ipi-install` chain, since that one is common to all platforms.

Unfortunately, adding the `ref: ipi-conf-aws-byo-iam-profile` before `chain: ipi-conf-aws` won't work because the step expects an install-config.yaml to already exist. So maybe we can include the step in the `ipi-conf-aws` chain and control when it's triggered via an env var?

Reply:
Another way is we can split the ipi-conf-aws chain to place these step in desired order in our install workflow.